File: myulib-core/src/main/java/com/myudog/myulib/internal/dynamics/VortexForceField.java
package com.myudog.myulib.internal.dynamics;

import com.myudog.myulib.api.core.dynamics.IForceField;
import net.minecraft.world.phys.Vec3;

public final class VortexForceField implements IForceField {
    private final Vec3 axis;

    public VortexForceField() {
        this(new Vec3(0.0, 1.0, 0.0));
    }

    public VortexForceField(Vec3 axis) {
        Vec3 normalized = axis == null ? new Vec3(0.0, 1.0, 0.0) : axis;
        double lengthSq = normalized.lengthSqr();
        this.axis = lengthSq < 0.0001 ? new Vec3(0.0, 1.0, 0.0) : normalized.normalize();
    }

    @Override
    public Vec3 calculateForce(Vec3 pos, Vec3 center, double strength) {
        Vec3 relativePos = pos.subtract(center);
        if (relativePos.lengthSqr() < 0.0001) {
            return Vec3.ZERO;
        }

        Vec3 tangent = relativePos.cross(axis);
        if (tangent.lengthSqr() < 0.0001) {
            return Vec3.ZERO;
        }

        return tangent.normalize().scale(strength);
    }
}

File: src/main/kotlin/com/myudog/myulib/internal/dynamics/VortexForceField.kt
package com.myudog.myulib.internal.dynamics

import com.myudog.myulib.api.dynamics.IForceField
import net.minecraft.util.math.Vec3d

internal class VortexForceField(val axis: Vec3d = Vec3d(0.0, 1.0, 0.0)) : IForceField {
    override fun calculateForce(pos: Vec3d, center: Vec3d, strength: Double): Vec3d {
        // 1. 取得從中心指向粒子的向量
        val relativePos = pos.subtract(center)
        if (relativePos.lengthSquared() < 0.001) return Vec3d.ZERO

        // 2. 計算切向向量 (外積)
        // 旋轉方向取決於軸的方向與外積順序
        val tangent = relativePos.crossProduct(axis).normalize()

        // 3. 渦流通常隨距離增加而減弱，或是保持恆定
        return tangent.multiply(strength)
    }
}